File: tam_agent/research/web_search.py
"""Web search integration for gathering market data."""

from __future__ import annotations

import httpx

from tam_agent.config import Settings
from tam_agent.models import SearchResult


class WebSearcher:
    """Executes web searches via Brave Search API, with graceful fallback."""

    def __init__(self, settings: Settings) -> None:
        self.api_key = settings.brave_search_api_key
        self.max_searches = settings.tam_agent_max_searches
        self._searches_done = 0

    @property
    def is_available(self) -> bool:
        return bool(self.api_key)

    @property
    def searches_remaining(self) -> int:
        return max(0, self.max_searches - self._searches_done)

    def search(self, query: str, num_results: int = 5) -> list[SearchResult]:
        """Execute a web search and return results.

        Returns an empty list if no API key is configured or quota is exhausted.
        """
        if not self.is_available or self.searches_remaining <= 0:
            return []

        self._searches_done += 1

        try:
            with httpx.Client(timeout=15.0) as client:
                resp = client.get(
                    "https://api.search.brave.com/res/v1/web/search",
                    headers={
                        "Accept": "application/json",
                        "Accept-Encoding": "gzip",
                        "X-Subscription-Token": self.api_key,
                    },
                    params={"q": query, "count": num_results},
                )
                resp.raise_for_status()
                data = resp.json()
        except (httpx.HTTPError, Exception):
            return []

        results: list[SearchResult] = []
        for item in data.get("web", {}).get("results", []):
            results.append(
                SearchResult(
                    title=item.get("title", ""),
                    url=item.get("url", ""),
                    snippet=item.get("description", ""),
                )
            )

        return results

    def multi_search(self, queries: list[str], num_results: int = 5) -> list[SearchResult]:
        """Execute multiple searches and deduplicate results."""
        all_results: list[SearchResult] = []
        seen_urls: set[str] = set()

        for query in queries:
            if self.searches_remaining <= 0:
                break
            for result in self.search(query, num_results):
                if result.url not in seen_urls:
                    seen_urls.add(result.url)
                    all_results.append(result)

        return all_results

File: src/research/web-search.ts
/** Web search integration for gathering market data. */

import type { Settings } from "../config.js";
import type { SearchResult } from "../models.js";

export class WebSearcher {
  private apiKey: string;
  private maxSearches: number;
  private searchesDone = 0;

  constructor(settings: Settings) {
    this.apiKey = settings.braveSearchApiKey;
    this.maxSearches = settings.tamAgentMaxSearches;
  }

  get isAvailable(): boolean {
    return this.apiKey.length > 0;
  }

  get searchesRemaining(): number {
    return Math.max(0, this.maxSearches - this.searchesDone);
  }

  async search(
    query: string,
    numResults = 5,
  ): Promise<SearchResult[]> {
    if (!this.isAvailable || this.searchesRemaining <= 0) {
      return [];
    }

    this.searchesDone++;

    try {
      const url = new URL("https://api.search.brave.com/res/v1/web/search");
      url.searchParams.set("q", query);
      url.searchParams.set("count", String(numResults));

      const resp = await fetch(url.toString(), {
        headers: {
          Accept: "application/json",
          "Accept-Encoding": "gzip",
          "X-Subscription-Token": this.apiKey,
        },
        signal: AbortSignal.timeout(15_000),
      });

      if (!resp.ok) return [];

      const data = (await resp.json()) as {
        web?: { results?: Array<{ title?: string; url?: string; description?: string }> };
      };

      return (data.web?.results ?? []).map((item) => ({
        title: item.title ?? "",
        url: item.url ?? "",
        snippet: item.description ?? "",
      }));
    } catch {
      return [];
    }
  }

  async multiSearch(
    queries: string[],
    numResults = 5,
  ): Promise<SearchResult[]> {
    const allResults: SearchResult[] = [];
    const seenUrls = new Set<string>();

    for (const query of queries) {
      if (this.searchesRemaining <= 0) break;
      const results = await this.search(query, numResults);
      for (const result of results) {
        if (!seenUrls.has(result.url)) {
          seenUrls.add(result.url);
          allResults.push(result);
        }
      }
    }

    return allResults;
  }
}
